pkg/grafana/datasources.go

```diff
@@ -0,0 +1,248 @@
+
+	// Configure Grafana client to use the correct organization
+	currentOrgID := s.grafanaClient.OrgID()
+	s.grafanaClient.WithOrgID(organization.ID)
+	defer s.grafanaClient.WithOrgID(currentOrgID)
+
+	// Fetch the currently configured datasources in Grafana
+	resp, err := s.grafanaClient.Datasources().GetDataSources()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get configured datasources: %w", err)
+	}
+
+	// Update or delete existing datasources
+	for _, currentDatasource := range resp.GetPayload() {
+		// Check if the current datasource exists in the desired datasources
+		index := slices.IndexFunc(desiredDatasources, func(d Datasource) bool {
+			return d.UID == currentDatasource.UID
+		})
+
+		if index >= 0 {
+			// Update the existing datasource
+			desiredDatasource := desiredDatasources[index]
+
+			logger.Info("updating datasource", "datasource", desiredDatasource.UID)
+			desiredDatasource, err = s.updateDatasource(ctx, desiredDatasource)
+			if err != nil {
+				return nil, err
+			}
+			// Set the ID of the updated datasource
+			desiredDatasources[index] = desiredDatasource
+			logger.Info("updated datasource", "datasource", desiredDatasource.UID)
+		}
+
+		if strings.HasPrefix(currentDatasource.UID, datasourceUIDPrefix) {
+			// Delete the datasource as it is no longer desired
+			logger.Info("deleting datasource", "datasource", currentDatasource.UID)
+			err := s.deleteDatasource(ctx, currentDatasource.UID)
+			if err != nil {
+				return nil, err
+			}
+			logger.Info("deleted datasource", "datasource", currentDatasource.UID)
+		}
+	}
+
+	// Create any new datasources that do not exist yet
+	for index := range desiredDatasources {
+		desiredDatasource := desiredDatasources[index]
+
+		// If the datasource ID is 0, it means it does not exist yet and needs to be created
+		// We already took care of updating existing datasources ID in the previous loop
+		if desiredDatasource.ID == 0 {
+			logger.Info("creating datasource", "datasource", desiredDatasource.UID)
+			desiredDatasource, err = s.createDatasource(ctx, desiredDatasource)
+			if err != nil {
+				return nil, err
+			}
+			logger.Info("datasource created", "datasource", desiredDatasource.UID)
+			// Set the ID of the created datasource
+			desiredDatasources[index] = desiredDatasource
+		}
+	}
+
+	return desiredDatasources, nil
+}
+
+// generateDatasources generates the list of datasources for a given organization.
+// It configures the datasources to use the appropriate multi-tenant headers based on the organization's tenant IDs.
+// It returns the list of desired datasources.
+func (s *Service) generateDatasources(ctx context.Context, organization Organization) (datasources []Datasource) {
+	// Multi-tenant header value is a pipe-separated list of tenant IDs
+	multiTenantIDsHeaderValue := strings.Join(organization.TenantIDs, "|")
+
+	// Add Loki datasource for logs only
+	datasources = append(datasources, DatasourceLoki().Merge(Datasource{
```

Comment:
Would it not make sense to have only 1 datasource that can manage alerts if we have 1 tenant and 1 for tenant if we have more than 1?